Full fastled replacement

DedeHai opened the pull request:
Completely removed dependency on fastled library for better color control and conversions. This also prevents accidentally pulling in fastled functions that were previously replaced (random() for example). 

- Copied all required functions and structs from fastled
- Optimized most of them for WLED use, many color manipulations in FX can now be replaced with native 32bit colors, potentially speeding things up with less color conversions
- Increased accuracy for some functions by scaling the fixedpoint math making use of native 32bit variables.
- Added a PRNG class for FX that rely on repatable random numbers
- Updated FX where to use the replacement functions where required

@netmindz Animartrix does not compile anymore since fastled.h is not available now. Can you please remove `#include <FastLED.h>` from ANIMartRIX.h? 0.16 compiles just fine without that include as well. I did not check older versions though, maybe some other measures are appropriate, I let you be the judge of that :)

DedeHai commented:
@willmmiles can you please review this? I had some issues with circular includes and I am not sure I did the includes of header in a good way.
To anyone taking a look at this: 
- feel free to examine the functions but generally focus on code structure/includes and if the way I split function declares in headers makes sense
- I tested and speed checked all relevant core functions and spent a lot of effort making them at least as fast as they were and ensured colors are correctly converted (hsv<->rgb mostly)
- I did not confirm all CRGB constructors/assignments do work but they are (or should be) straight copies. the ones I modified, I checked
- any bugs that are in color functions should be minor and can be worked out as they are discovered

netmindz commented:
Updated AnimARTrix - e6b7509b391cab97b59b8eb19580e67d4203ae7d

TripleWhy commented on wled00/colors.cpp:
```diff
@@ -1,8 +1,17 @@
+  - direct conversion (assignment or construction) from CHSV/CHSV32 to CRGB/CRGBW use the "rainbow" method (nicer colors)
+  - converting CRGB(W) to HSV32 color is quite accurate but still not 100% (but much more accurate than fastled's "hsv2rgb_approximate" function)
+  - when converting CRGB(W) to HSV32 and back, "hsv2rgb_spectrum" preserves the colors better than the _rainbow version
+  - to manipulate an RGB color in HSV space, use the adjust_color function or the CRGBW.adjust_hue method
```
I think I'd expect this comment in the header

DedeHai replied:
I have no preference, since the structs are in the header, you have a point.

TripleWhy commented on wled00/colors.cpp:
```diff
@@ -38,13 +46,9 @@ uint32_t color_add(uint32_t c1, uint32_t c2, bool preserveCR)
 
   if (preserveCR) { // preserve color ratios
-    uint32_t max = std::max(r,g); // check for overflow note
+    uint32_t max = std::max(r,g); // check for overflow
```
well this line doesn't check anything. I'd move the comment up one line (or down to the if)

TripleWhy commented on wled00/colors.cpp:
```diff
@@ -86,8 +88,24 @@ uint32_t color_fade(uint32_t c1, uint8_t amount, bool video)
+  note: inputs are 32bit to speed up the function, useful input value ranges are -255 to +255
+  note2: if only one hue change is needed, use CRGBW.adjust_hue() instead (much faster)
+*/
+__attribute__((optimize("O3"))) void adjust_color(CRGBW& rgb, int32_t hueShift, int32_t valueChange, int32_t satChange) {
```
* I mean... CRGBW is WLED code now. You can just make this function a member, too.
* I don't see this being used anywhere? Do we need it?

DedeHai replied:
CRGBW was always WLED code. This was created in an old PR (copy segment) that may see a revival soon. Can move it later if its still suitable.

TripleWhy commented on wled00/colors.cpp:
```diff
@@ -232,64 +250,205 @@ CRGBPalette16 generateRandomPalette()  // generate fully random palette
-      rgb = v << 16 | v << 8 | v;
-      return;
+// convert HSV (16bit hue) to RGB (32bit with white = 0), optimized for speed
+__attribute__((optimize("O3"))) void hsv2rgb_spectrum(const CHSV32& hsv, CRGBW& rgb) {
```
the `w` value is never modified. you should probably compute it, zero it, or use `CRGB` instead

TripleWhy replied:
actually, why is CRGBW not the return type? do you intentionally keep w as is? Even then, I think it would be better as a return value.

DedeHai replied:
yes, w is intentionally not modified. in place manipulation is generally faster than copy and return copy.

TripleWhy replied:
generally faster for complex types, yes. but this is a 32 bit int

DedeHai replied:
in code that has to be fast, every instruction counts.

TripleWhy replied:
did you actually compare different versions on instruction level? more complex function signatures can cost performance, too. it can also also be slower to use references instead of copies if the type isn't very wide. Not saying that any of this applies necessarily here, just that stuff isn't as obvious as it looks, especially if you also have an optimizer. So if you *really* want to be fast, you *always* need to measure

DedeHai replied:
you are right, and yes I did benchmark it at one point.

TripleWhy commented on wled00/colors.cpp:
```diff
@@ -232,64 +250,205 @@ CRGBPalette16 generateRandomPalette()  // generate fully random palette
-    int32_t r = (rgb>>16)&0xFF;
-    int32_t g = (rgb>>8)&0xFF;
-    int32_t b = rgb&0xFF;
+// CHSV to CRGB, dumb conversion: slower so this should not be used in time critical code, use rainbow version instead
```
There are a lot of color space conversion functions now. from a quick glance, I can't tell what their differences are. I would love a short summary of which function to use when to be added to all conversion function comments, and, if necessary, a more detailed comparison comment block before the first one or at the top of the header. If a short summary is too long for all of them, at least point to the detailed comparison

DedeHai replied:
there is a comment block that explains it a bit but basically there is only two flavours of rgb2hsv: spectrum and rainbow and those are nicely explained in fastled wiki. 
all other conversions are more or less self explaining. so anything specific you want explained?

TripleWhy replied:
See above. If you believe an explanation in another location is sufficient and will forever be available, at least link to it so that a reader gets a chance to find it without already knowing about it

TripleWhy commented on wled00/colors.cpp:
```diff
@@ -232,64 +250,205 @@ CRGBPalette16 generateRandomPalette()  // generate fully random palette
-    if (hsv.s == 0)  return; // gray value
+    hsv.s = delta == maxval ? 255 : (255 * delta) / maxval; // faster on fully saturated colors, slightly slower otherwise
+    //hsv.s = (255 * delta) / maxval;
+    //if (hsv.s == 0)  return; // gray value // assuming gray values are passed rarely, this can be omitted to increase speed
```
remove

TripleWhy commented on wled00/colors.h:
```diff
@@ -0,0 +1,910 @@
+// it can be useful to avoid back and forth conversions between uint32_t and fastled CRGB
+struct CRGBW {
+  union {
+    uint32_t color32; // Access as a 32-bit value (0xWWRRGGBB)
```
I believe this technically violates strict aliasing and is undefined behavior, but whatever ^^

DedeHai replied:
it is a bit problematic (ambiguous conversions), but it allows direct comparison, constructor etc. from 32bit colors and speed is key in colors as its done for every pixel. Is there a better way that ensures speed?

TripleWhy replied:
c++20 has `std::bit_cast`, gcc has `-fno-strict-aliasing` (which may be applied selectively per compilation unit), and there is `memcpy` which may be optimized out. C would also allow this code without changes.

DedeHai replied:
from what I understand: a union does not violate strict aliasing.

TripleWhy replied:
not simply by existing, no. but it does if you read a value that isn't the last one you wrote

DedeHai replied:
do you have an example for what you mean? and how is it not an issue in fastled?

TripleWhy replied:
example from https://en.cppreference.com/w/cpp/language/union:

```c++
union S
{
    std::int32_t n;     // occupies 4 bytes
    std::uint16_t s[2]; // occupies 4 bytes
    std::uint8_t c;     // occupies 1 byte
};                      // the whole union occupies 4 bytes
 
int main()
{
    S s = {0x12345678}; // initializes the first member, s.n is now the active member
    // At this point, reading from s.s or s.c is undefined behavior,
    // but most compilers define it.
    std::cout << std::hex << "s.n = " << s.n << '\n';
 
    s.s[0] = 0x0011; // s.s is now the active member
    // At this point, reading from s.n or s.c is undefined behavior,
    // but most compilers define it.
    std::cout << "s.c is now " << +s.c << '\n' // 11 or 00, depending on platform
              << "s.n is now " << s.n << '\n'; // 12340011 or 00115678
}
```

In `CRGB` this isn't an issue because it defines 3 ways to access each channel, but all of them access the same memory and are of the same type. In `CRGBW` you alias `uint32_t` with `uint8_t[4]`. Which is _probably_ fine, but if you access one after writing the other, it's technically UB. As noted in the above example, the compiler _may_ define this to be fine, but I couldn't find any official statement for gcc. It may therefore even change from version to version.

DedeHai replied:
ok, thanks for the explanation. I just read that byte access is exempt from strict aliasing, so should really not be an issue.

TripleWhy commented on wled00/ir.cpp:
```diff
@@ -128,22 +129,20 @@ static void changeEffectSpeed(int8_t amount)
-    prim_hsv.h = (byte)new_val;
-    hsv2rgb_rainbow(prim_hsv, fastled_col);
+    CHSV32 prim_hsv = CRGBW(sseg.colors[0]);
+    prim_hsv.h += (amount<<8);
```
Isn't that what adjust_hue is for?

DedeHai replied:
I think yes, I updated files out of order, this was updated before I added the adjust_hue. will check.

TripleWhy commented on wled00/wled.cpp:
```diff
@@ -513,10 +513,6 @@ void WLED::setup()
 
-  // Seed FastLED random functions with an esp random value, which already works properly at this point.
-  const uint32_t seed32 = hw_random();
-  random16_set_seed((uint16_t)seed32);
```
I don't think you restore this behavior in another location, do you?

DedeHai replied:
no, I removed this under the impression that all FX do their own seed now, but that could be wrong. will check if this is still required.

DedeHai commented on wled00/colors.h:
```diff
@@ -0,0 +1,910 @@
+#ifndef WLED_COLORS_H
+#define WLED_COLORS_H
+
+// note: some functions/structs have been copied from fastled library, modified and optimized for WLED
```
that is a valid point, could try to find the original author of the functions copied and ask.
as for the re-licensing: AFAIK all major contributors were contacted and asked for permission.

TripleWhy replied:
licensing is all about technicality. And technically, you need permission from *ever single* contributor whose contributions were not removed *prior* to the license change. And I know for a fact that you didn't ask everyone.

While I think it's unlikely that comes out of this, if someone *wants* to cause trouble because of this, you expose a lot of attack surface with this move

DedeHai commented on wled00/colors.h:
```diff
@@ -0,0 +1,910 @@
+#define WLED_COLORS_H
+
+// note: some functions/structs have been copied from fastled library, modified and optimized for WLED
+
```
it is FastLED @ 3.6.0, could include that in the comment

DedeHai commented on wled00/fcn_declare.h:
```diff
@@ -600,6 +506,30 @@ uint32_t sqrt32_bw(uint32_t x);
 */
+
+// PRNG for 16bit and 8bit random numbers used by some effects (fastled replacement)
+class PRNG {
```
I am open to suggestions. I did not want to alter the WLED file structure. What would be a better place to put this?

TripleWhy replied:
PRGN.h

DedeHai replied:
create a new file for 20 lines of code?

TripleWhy replied:
yes. would you really consider the current state of things a good one? ^^

DedeHai commented on wled00/FX.cpp:
```diff
@@ -68,17 +69,14 @@
 //#define MAX_FREQUENCY   5120
 //#define MAX_FREQ_LOG10  3.71f
 
+PRNG prng; // pseudo-random number generator class
```
you are probably right, I am not familiar with the use of namespaces

TripleWhy replied:
this is an unnamed namespace, which achieves the same thing as `static PRNG prng;`: it makes this token visible only to the current translation unit (the cpp file), instead of globally. why unnamed namespace instead of static? It's the newer version of doing this. is it any better? I don't know.

DedeHai replied:
moving PRNG to its own file, `prng` is now only in global scope of FX.cpp so this is not required.

TripleWhy replied:
Moving the class is unrelated to moving this variable

DedeHai replied:
so why exacly does this one require a namespace but any other global variable does not?

TripleWhy replied:
ah sorry, I didn't notice this was a public global variable, declared elsewhere. I was assuming this was a local variable used in FX.cpp only.  
As I said, the namespace thing is the same as `static`. So if you need this outside of FX.cpp, keep it global (and maybe put the definition in the prng.cpp, if you moved the declaration to prng.h)

DedeHai replied:
since its not used anywhere else than FX.cpp, it would make sense to declare it just there no? does it then HAVE TO be static or is that just best practice to avoid compiler issues?

TripleWhy replied:
It doesn't *have* to be static, but that limits the visibility to the cpp file. You really don't want to declare publicly visible stuff in cpp files, because you don't usually want to include cpp files to make them available. Also not polluting the global namespace is always a good idea :)

DedeHai commented on wled00/FX.cpp:
```diff
@@ -1759,30 +1757,30 @@ static const char _data_FX_MODE_MULTI_COMET[] PROGMEM = "Multi Comet@!,Fade;!,!;
-    SEGENV.step = RGBW32(random8(), random8(), random8(), 0);
-    SEGENV.aux0 = random16();
+    SEGENV.step = RGBW32(prng.random8(), prng.random8(), prng.random8(), 0);
+    SEGENV.aux0 = prng.random16();
```
not the way the FX works, see seed retention below. the idea is to be able to have two segments in sync.

TripleWhy replied:
yes, and that works with the prng usage blelow. but if we want to also synchronize the initial state, we either need to synchroinze step and aux anyway, or we need to make sure prng's seed is synchronized at this point in time, and I don't think either happens.

DedeHai replied:
I am not sure how (and why) PRNG is used here and I don't want to introduce new bugs so better leave it as-is.

DedeHai commented on wled00/FX.cpp:
```diff
@@ -1759,30 +1757,30 @@ static const char _data_FX_MODE_MULTI_COMET[] PROGMEM = "Multi Comet@!,Fade;!,!;
+    SEGENV.aux0 = prng.random16();
   }
-  unsigned prevSeed = random16_get_seed(); // save seed so we can restore it at the end of the function
+  unsigned prevSeed = prng.getSeed(); // save seed so we can restore it at the end of the function
```
no, see my comment above: segment sync. this may change in the future though (if copy segment will become a thing)

TripleWhy replied:
I fail to see the relevance of the response to my comment

DedeHai replied:
If each effect has its own, they do not sync.

DedeHai commented on wled00/FX.cpp:
```diff
@@ -2618,14 +2616,14 @@ static uint16_t twinklefox_base(bool cat)
   unsigned bglight = bg.getAverageLight();
   if (bglight > 64) {
-    bg.nscale8_video(16); // very bright, so scale to 1/16th
+    bg = color_fade(bg, 16, true); // very bright, so scale to 1/16th
```
I think so, will check again.

DedeHai replied:
it is equivalent. nscale8_video is an original fastled function for CRGB.

DedeHai commented on wled00/FX.cpp:
```diff
@@ -4146,17 +4144,17 @@ static const char _data_FX_MODE_PHASEDNOISE[] PROGMEM = "Phased Noise@!,!;!,!;!"
 uint16_t mode_twinkleup(void) {                 // A very short twinkle routine with fade-in and dual controls. By Andrew Tuline.
-  unsigned prevSeed = random16_get_seed();      // save seed so we can restore it at the end of the function
-  random16_set_seed(535);                       // The randomizer needs to be re-set each time through the loop in order for the same 'random' numbers to be the same each time through.
+  unsigned prevSeed = prng.getSeed();      // save seed so we can restore it at the end of the function
```
thanks for pointing this out, messy code alignment eludes my eyes ;)

DedeHai commented on wled00/FX.cpp:
```diff
@@ -5696,8 +5695,7 @@ uint16_t mode_2DSunradiation(void) {                   // By: ldirko https://edi
     for (int i = 0; i < (cols + 2); i++) {
-      //byte col = (inoise8_raw(i * someVal, j * someVal, t)) / 2;
-      byte col = ((int16_t)perlin8(i * someVal, j * someVal, t) - 0x7F) / 3;
+      byte col = ((int16_t)perlin8(i * someVal, j * someVal, t) - 127) >> 2; // about +/- 32
```
the decimal is more readable to me, personal preference. the shift is actually a bugfix from an earlier commit to make it more in-line with the original inoise8_raw(). I need to check again but I remember doing this intentionally to get the original look.
shifts are indeed less readable, but the compiler does not always convert a division to a shift so I tend to use them explicitly.

TripleWhy replied:
how sure are you about that not working reliably though? I believe that is an ancient myth

DedeHai replied:
not a myth. in this case, it would not replace a /4 by a shift and rightfully so.

TripleWhy replied:
why is that? because the left hand argument is a signed integer? well in that case using right shifts isn't exactly a great idea anyway

DedeHai replied:
yes, because it is signed. not great, not terrible either, but fast. in this case: good enough.

TripleWhy replied:
maybe not terrible, but undefined behavior.

also note that this is a risc cpu. as long as there is a single assembly instruction for integer division, it should take the same time as any other assembly instruction, including a bit shift (at least in theory)

DedeHai commented on wled00/FX.cpp:
```diff
@@ -6013,7 +6011,7 @@ uint16_t mode_2Dfloatingblobs(void) {
   // Bounce balls around
   for (size_t i = 0; i < Amount; i++) {
-    if (SEGENV.step < strip.now) blob->color[i] = add8(blob->color[i], 4); // slowly change color
+    if (SEGENV.step < strip.now) blob->color[i] = blob->color[i] + 4; // slowly change color
```
I have no idea about the original... my thought exactly :)
good point about the +=

DedeHai commented on wled00/FX.cpp:
```diff
@@ -10050,7 +10048,7 @@ uint16_t mode_particle1Dsonicstream(void) {
 
   loudness = fftResult[baseBin];// + fftResult[baseBin + 1];
-  int mids = sqrt16((int)fftResult[5] + (int)fftResult[6] + (int)fftResult[7] + (int)fftResult[8] + (int)fftResult[9] + (int)fftResult[10]); // average the mids, bin 5 is ~500Hz, bin 10 is ~2kHz (see audio_reactive.h)
+  int mids = sqrt32_bw((int)fftResult[5] + (int)fftResult[6] + (int)fftResult[7] + (int)fftResult[8] + (int)fftResult[9] + (int)fftResult[10]); // average the mids, bin 5 is ~500Hz, bin 10 is ~2kHz (see audio_reactive.h)
```
speed optimization. and yes, the 6 8bit values will fit an int32 ;)

TripleWhy replied:
I was talking about the return value. `sqrt16` returned a `uint8_t`. `sqrt32_bw` returns a `uint32_t`

DedeHai commented on wled00/FXparticleSystem.cpp:
```diff
@@ -648,21 +648,18 @@ void ParticleSystem2D::ParticleSys_render() {
       }
     }
-    renderParticle(i, brightness, baseRGB, particlesettings.wrapX, particlesettings.wrapY);
+    renderParticle(i, brightness, CRGB(baseRGB), particlesettings.wrapX, particlesettings.wrapY);
```
yes, this will change in a bit: the whole PS will go 32bit colors.

RJac702 commented on wled00/FX.h:
```diff
@@ -42,10 +38,12 @@
 #ifndef RGBW32
 #define RGBW32(r,g,b,w) (uint32_t((byte(w) << 24) | (byte(r) << 16) | (byte(g) << 8) | (byte(b))))
 #endif
+**/
```
ok

DedeHai commented:
@Aircoookie any thoughts on how I fixed the licensing issue?
- moved some of the FastLED stuff into its own folder, adding MIT license there
- referenced that license in all places where FastLED functions/palettes were added within WLED code

netmindz commented:
Just to add something else into the mix, have you been keeping an eye on the recent FastLED releases? There are some new features like WaveFX which could be useful to write new effects.
It's moving much more from just being led output plus a few utilities to a wider effects platform

DedeHai commented on usermods/audioreactive/audio_reactive.cpp:
```diff
@@ -1994,12 +1994,12 @@ CRGB AudioReactive::getCRGBForBand(int x, int pal) {
     case 2:
       b = map(x, 0, 255, 0, NUM_GEQ_CHANNELS/2); // convert palette position to lower half of freq band
       hsv = CHSV(fftResult[b], 255, x);
```
that sure can be arranged. :)

DedeHai commented on wled00/FXparticleSystem.cpp:
```diff
@@ -595,19 +595,17 @@ void ParticleSystem2D::render() {
+        CHSV32 baseHSV = baseRGB;
         baseHSV.s = min(baseHSV.s, particles[i].sat); // set the saturation but don't increase it
-        hsv2rgb(baseHSV, baseRGB.color32); // convert back to RGB
+        hsv2rgb_spectrum(baseHSV, baseRGB); // convert back to RGB
```
this is intentional, particle system does not use nor handle white channel
